core/src/main/java/io/substrait/relation/Set.java

```diff
@@ -42,7 +46,88 @@ public static Set.SetOp fromProto(SetRel.SetOp proto) {
+      throw new IllegalArgumentException("Set's input records have different number of fields");
+    }
+
+    // As defined in https://substrait.io/relations/logical_relations/#set-operation-types
+    return switch (getSetOp()) {
+      case UNKNOWN -> first; // alternative would be to throw an exception
+      case MINUS_PRIMARY, MINUS_PRIMARY_ALL, MINUS_MULTISET -> first;
+      case INTERSECTION_PRIMARY -> coalesceNullabilityIntersectionPrimary(first, rest);
+      case INTERSECTION_MULTISET, INTERSECTION_MULTISET_ALL -> coalesceNullabilityIntersection(
+          first, rest);
+      case UNION_DISTINCT, UNION_ALL -> coalesceNullabilityUnion(first, rest);
+    };
+  }
+
+  /** If field is nullable in any of the inputs, it's nullable in the output */
+  private Type.Struct coalesceNullabilityUnion(Type.Struct first, List<Type.Struct> rest) {
+
+    List<Type> fields = new ArrayList<>();
+    for (int i = 0; i < first.fields().size(); i++) {
+      Type typeA = first.fields().get(i);
+      int finalI = i;
+      fields.add(
+          rest.stream()
+              .map(struct -> struct.fields().get(finalI))
+              .reduce(
+                  typeA, (curr, next) -> next.nullable() ? TypeCreator.asNullable(curr) : curr));
+    }
+    return Type.Struct.builder().fields(fields).nullable(first.nullable()).build();
+  }
+
+  /**
+   * If field is nullable in primary input and in any of the other of the inputs, it's nullable in
+   * the output
+   */
+  private Type.Struct coalesceNullabilityIntersectionPrimary(
+      Type.Struct first, List<Type.Struct> rest) {
+
+    List<Type> fields = new ArrayList<>();
+    for (int i = 0; i < first.fields().size(); i++) {
+      Type typeA = first.fields().get(i);
+      if (!typeA.nullable()) {
+        // Just to make this case explicit and to short-circuit, logic below would work without too
+        fields.add(typeA);
+        continue;
+      }
+      int finalI = i;
+      boolean anyOtherIsNullable = rest.stream().anyMatch(t -> t.fields().get(finalI).nullable());
+      fields.add(anyOtherIsNullable ? typeA : TypeCreator.asNotNullable(typeA));
+    }
+    return Type.Struct.builder().fields(fields).nullable(first.nullable()).build();
+  }
+
+  /** If field is required in any of the inputs, it's required in the output */
+  private Type.Struct coalesceNullabilityIntersection(Type.Struct first, List<Type.Struct> rest) {
+    List<Type> fields = new ArrayList<>();
+    for (int i = 0; i < first.fields().size(); i++) {
+      Type typeA = first.fields().get(i);
+      if (!typeA.nullable()) {
+        // Just to make this case explicit and to short-circuit, logic below would work without too
+        fields.add(typeA);
+        continue;
+      }
+      int finalI = i;
+      boolean anyOtherIsRequired = rest.stream().anyMatch(t -> !t.fields().get(finalI).nullable());
+      fields.add(anyOtherIsRequired ? TypeCreator.asNotNullable(typeA) : typeA);
+    }
```

Comment:
Looks good.  Tiny nit - both of these functions share a lot of common code (only 2 lines differ) - could they be refactored somehow?  Not important though :)

Reply:
I had it before as a single function, but it ends up making it harder to read since while the difference is small, it's non-trivial :/